src/librustc_typeck/check/mod.rs

```diff
@@ -1632,47 +1632,56 @@ fn check_opaque_for_inheriting_lifetimes(tcx: TyCtxt<'tcx>, def_id: LocalDefId,
     struct ProhibitOpaqueVisitor<'tcx> {
         opaque_identity_ty: Ty<'tcx>,
         generics: &'tcx ty::Generics,
+        ty: Option<Ty<'tcx>>,
     };
 
     impl<'tcx> ty::fold::TypeVisitor<'tcx> for ProhibitOpaqueVisitor<'tcx> {
         fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
             debug!("check_opaque_for_inheriting_lifetimes: (visit_ty) t={:?}", t);
+            self.ty = Some(t);
             if t == self.opaque_identity_ty { false } else { t.super_visit_with(self) }
```

Comment:
```suggestion
            if t != self.opaque_identity_ty && t.super_visit_with(self) {
                self.ty = Some(t);
                return true;
            }
            false
```

This avoids "resetting" `self.ty` in `visit_region`.